os/exec/process_monitor.go

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"os"
+	oexec "os/exec"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+var (
+	defaultStdoutSuffix         = "out"
+	defaultStderrSuffix         = "err"
+	defaultProcessCheckInterval = time.Second
+)
+
+var (
+	errUnableToStartClosed        = fmt.Errorf("unable to start: process monitor Closed()")
+	errUnableToStartRunning       = fmt.Errorf("unable to start: process already running")
+	errUnableToStopClosed         = fmt.Errorf("unable to stop: process monitor Closed()")
+	errUnableToStopStoped         = fmt.Errorf("unable to stop: process not running")
+	errUnableToClose              = fmt.Errorf("unable to close: process monitor already Closed()")
+	errPathNotSet                 = fmt.Errorf("cmd: no path specified")
+	errArgsRequiresPathAsFirstArg = fmt.Errorf("cmd: args[0] un-equal to path")
+)
+
+func (c Cmd) validate() error {
```

Comment:
might be better to move these 2 functions to types.go to be closer with their struct definition?